pkg/hub/manager.go

```diff
@@ -0,0 +1,78 @@
+package hub
+
+import (
+	"context"
+	"time"
+
+	clusterv1client "github.com/open-cluster-management/api/client/cluster/clientset/versioned"
+	clusterv1informers "github.com/open-cluster-management/api/client/cluster/informers/externalversions"
+	workclientset "github.com/open-cluster-management/api/client/work/clientset/versioned"
+	workinformers "github.com/open-cluster-management/api/client/work/informers/externalversions"
+	"github.com/openshift/library-go/pkg/controller/controllercmd"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/dynamic/dynamicinformer"
+
+	"github.com/open-cluster-management/work/pkg/hub/controllers/finalizercontroller"
+)
+
+// RunControllerManager starts the controllers on hub to manage manifest works.
+func RunControllerManager(ctx context.Context, controllerContext *controllercmd.ControllerContext) error {
+	clusterClient, err := clusterv1client.NewForConfig(controllerContext.KubeConfig)
+	if err != nil {
+		return err
+	}
+	clusterInformers := clusterv1informers.NewSharedInformerFactory(clusterClient, 5*time.Minute)
+	clusterLister := clusterInformers.Cluster().V1().SpokeClusters().Lister()
+
+	workClient, err := workclientset.NewForConfig(controllerContext.KubeConfig)
+	if err != nil {
+		return err
+	}
+	workInformers := workinformers.NewSharedInformerFactory(workClient, 5*time.Minute)
+	workLister := workInformers.Work().V1().ManifestWorks().Lister()
+
+	dynamicClient, err := dynamic.NewForConfig(controllerContext.KubeConfig)
+	if err != nil {
+		return err
+	}
+	dynamicInformers := dynamicinformer.NewDynamicSharedInformerFactory(dynamicClient, 5*time.Second)
+
+	// set up controllers to add and handle finalizer on role
+	roleGVR := schema.GroupVersionResource{
+		Group:    "rbac.authorization.k8s.io",
+		Version:  "v1",
+		Resource: "roles",
+	}
+	roleInformer := dynamicInformers.ForResource(roleGVR)
+	addFinalizerControllerForRole := finalizercontroller.NewAddFinalizerController(controllerContext.EventRecorder, roleGVR,
+		roleInformer, clusterLister, dynamicClient, "AddFinalizerControllerForRole")
+
+	finalizeControllerForRole := finalizercontroller.NewFinalizeController(controllerContext.EventRecorder, roleGVR,
+		roleInformer, clusterLister, workClient.WorkV1(), workLister, dynamicClient, "FinalizeControllerForRole")
+
+	// set up controllers to add and handle finalizer on rolebinding
+	roleBindingGVR := schema.GroupVersionResource{
+		Group:    "rbac.authorization.k8s.io",
+		Version:  "v1",
+		Resource: "rolebindings",
+	}
+	roleBindingInformer := dynamicInformers.ForResource(roleBindingGVR)
+	addFinalizerControllerForRoleBinding := finalizercontroller.NewAddFinalizerController(controllerContext.EventRecorder, roleBindingGVR,
+		roleBindingInformer, clusterLister, dynamicClient, "AddFinalizerControllerForRoleBinding")
+
+	finalizeControllerForRoleBinding := finalizercontroller.NewFinalizeController(controllerContext.EventRecorder, roleBindingGVR,
```

Comment:
i see the motivation to use dynamicClient...But I think we could have one addFinalizerController to update both finalizer of role/rolebinding, and another finalizerController to remove finalizers.